### PR TITLE
DeepSpeed Revamp

### PR DESCRIPTION
**Pull request by pacman100**
### What does this PR do?
1. Support for DeepSpeed Config File
2. Support for gradient clipping and offloading params w/o config file
3. Fixing scheduler w/o config file
4. Simplifying model, optimizer and scheduler wrappers by removing stale code of current wrappers and directly relying on DeepSpeed.
5. Using HFDeepSpeedConfig when in Zero Stage-3 as per user consent to handle Deepspeed ZeRO-3 param gathering and automatically splitting the model onto multiple gpus during `from_pretrained` call

ToDo:
- [x] Write Tests
- [x] Redo experiments from [Testing Trainer and Accelerate Integration of DeepSpeed](https://wandb.ai/smangrul/Accelerate-MRPC-DeepSpeed/reports/Testing-Trainer-and-Accelerate-Integrations-of-DeepSpeed--VmlldzoyMDMwNzE1?accessToken=xz2scw676qasp8xta96ol11kcodxygg4dmploganrnyj2167t1vratbp12wlt8v2) to see if gaps between Accelerate integration and Trainer integrations are fixed.
- [x] ZeRo Stage-3 inference
- [x] Adding example leveraging config-file and related saving and loading of model when using deepspeed.
- [x] Documentation

### CODE REVIEW

File: src/accelerate/accelerator.py [hunk 1]
```diff
@@ -499,81 +527,116 @@ def _prepare_deepspeed(self, *args):
...
-            self.deepspeed_engine = engine  # pointing for deepspeed_engine.backward()
+                elif isinstance(result[i], (torch.optim.Optimizer, DummyOptim)):
+                    result[i] = optimizer
+                elif (isinstance(result[i], (torch.optim.lr_scheduler._LRScheduler, DummyScheduler))) or (
+                    type(result[i]).__name__ in deepspeed.runtime.lr_schedules.VALID_LR_SCHEDULES
+                ):
+                    result[i] = scheduler
+            self.deepspeed_engine_wrapped = DeepSpeedEngineWrapper(
+                engine
+            )  # pointing for deepspeed_engine_wrapped.backward()
```
Comment (sgugger) [inline src/accelerate/accelerator.py:628]: Put the comment above so everything fits on one line.

File: src/accelerate/accelerator.py [hunk 2]
```diff
@@ -499,81 +527,116 @@ def _prepare_deepspeed(self, *args):
...
             assert (
                 len(self._models) == 1
             ), "You can't use same `Accelerator()` instance with 2 models when using DeepSpeed"
 
         if self.distributed_type == DistributedType.DEEPSPEED:
             assert hasattr(
-                self, "deepspeed_engine"
-            ), "You need to pass the model along the optimizer when using Deepspeed."
+                self, "deepspeed_engine_wrapped"
+            ), "You need to pass the model along with the optimizer when using Deepspeed."
```
Comment (sgugger) [inline src/accelerate/accelerator.py:639]: Let's take this PR as an opportunity to remove those assers and use proper exceptions :-)

File: src/accelerate/commands/config/cluster.py [hunk 3]
```diff
@@ -77,24 +77,72 @@ def get_cluster_input():
...
+                default=False,
+                error_message="Please enter yes or no.",
             )
-
-            if deepspeed_config["zero_stage"] >= 2:
-                deepspeed_config["offload_optimizer_device"] = _ask_field(
-                    "Where to offload optimizer states? [NONE/cpu/nvme]: ",
+            if use_deepspeed_config:
+                deepspeed_config["deepspeed_config_file"] = _ask_field(
+                    "Please enter filepath of the DeepSpeed json config file: ",
```
Comment (sgugger) [inline src/accelerate/commands/config/cluster.py:88]:
```suggestion
                    "Please enter the path to the json DeepSpeed config file: ",
```

File: src/accelerate/accelerator.py [hunk 4]
```diff
@@ -648,13 +711,17 @@ def clip_grad_norm_(self, parameters, max_norm, norm_type=2):
                 if parameters == [p for p in model.parameters()]:
                     model.clip_grad_norm_(max_norm, norm_type)
                     return
+        elif self.distributed_type == DistributedType.DEEPSPEED:
+            return
```
Comment (sgugger) [inline src/accelerate/accelerator.py:715]: Maybe put a comment to explain DeepSpeed will handle this?

File: src/accelerate/accelerator.py [hunk 5]
```diff
@@ -648,13 +711,17 @@ def clip_grad_norm_(self, parameters, max_norm, norm_type=2):
...
         self.unscale_gradients()
         torch.nn.utils.clip_grad_norm_(parameters, max_norm, norm_type=norm_type)
 
     def clip_grad_value_(self, parameters, clip_value):
         """
         Should be used in place of `torch.nn.utils.clip_grad_value_`.
         """
+        if self.distributed_type in [DistributedType.DEEPSPEED, DistributedType.FSDP]:
+            return
```
Comment (sgugger) [inline src/accelerate/accelerator.py:724]: Same here.

File: src/accelerate/commands/launch.py [hunk 6]
```diff
@@ -57,6 +58,56 @@ def launch_command_parser(subparsers=None):
         action="store_true",
         help="Whether to use deepspeed.",
     )
+    parser.add_argument(
+        "--deepspeed_config_file",
+        default=None,
+        type=str,
+        help="DeepSpeed's config file (useful only when `use_deepspeed` flag is passed).",
+    )
+    parser.add_argument(
```
Comment (sgugger) [inline src/accelerate/commands/launch.py:67]: Not really sure it's worth adding all of those below. The DeepSpeed config is enough IMO.

File: src/accelerate/utils/deepspeed.py [hunk 7]
```diff
@@ -12,58 +12,26 @@
...
 
-        # this will ensure deepspeed gradient_accumulation matches user's accumulation
-        self.micro_steps += 1
+        # gradient accumulation check
+        # gradient clipping
+        # optimizer step
+        # zero grad
+        # checking overflow
+        # lr_scheduler step
+        self.engine.step()
```
Comment (sgugger) [inline src/accelerate/utils/deepspeed.py:34]: Not sure I understand the comments. Is it everything that happens in this step? If that's the case, can you add a comment stating it at the top?

File: tests/deepspeed/test_deepspeed.py [hunk 8]
```diff
@@ -0,0 +1,546 @@
...
+        with io.open(self.ds_config_file[ZERO3], "r", encoding="utf-8") as f:
+            config_zero3 = json.load(f)
+            # The following setting slows things down, so don't enable it by default unless needed by a test.
+            # It's in the file as a demo for users since we want everything to work out of the box even if slower.
+            config_zero3["zero_optimization"]["stage3_gather_16bit_weights_on_model_save"] = False
+
+        self.ds_config_dict = dict(
+            zero2=config_zero2,
+            zero3=config_zero3,
+        )
```
Comment (sgugger) [inline tests/deepspeed/test_deepspeed.py:114]:
```suggestion
        self.ds_config_dict = dict(zero2=config_zero2, zero3=config_zero3)
```
Fits in one line ;-)

**Review by sgugger — APPROVED** [review]
Thanks a lot for all the work and very nice new tests!
Can you confirm whether this is fully backward compatible, or if there are any breaking changes, could you document them?

File: src/accelerate/accelerator.py [hunk 9]
```diff
@@ -163,6 +172,25 @@ def __init__(
...
+                )
+            if compare_versions("deepspeed", "<", "0.6.4"):
+                raise ImportError("DeepSpeed version must be >= 0.6.4. Please update DeepSpeed.")
+            if os.environ.get("DEEPSPEED_ZERO3_INIT", "false") == "true" or deepspeed_plugin.zero3_init_flag:
+                if not is_transformers_available():
+                    raise Exception(
+                        "When `zero3_init_flag` is set, it requires Transformers to be installed. "
+                        "Please run `pip3 install transformers`."
+                    )
+                from transformers.deepspeed import HfDeepSpeedConfig
```
[the remark this answers is not in the document]
Comment (pacman100) [reply src/accelerate/accelerator.py:186]: I thought about this. The thing is a weakref to `HFDeepSpeedConfig` is created in that file (transformers.deepspeed). This is important only when using ZeRO Stage-3 when we don't want to load Transformer models fully on CPU/GPU and we want to directly partition the model parameters across GPUs. This weakref `_hf_deepspeed_config_weak_ref` is used in transformers `modeling_utils.py` to check if deepspeed zero stage3 is enabled. If it is enabled DeepSpeed functionality of `zero.init` is used to directly partition model parameters across GPUs. It is used by `from_config` (when training from scratch) and `from_pretrained` (when finetuning) methods. 

Snippet in `modeling_utils.py`:
```python
if is_deepspeed_zero3_enabled():
            import deepspeed
  
            logger.info("Detected DeepSpeed ZeRO-3: activating zero.init() for this model")
            # this immediately partitions the model across all gpus, to avoid the overhead in time
            # and memory copying it on CPU or each GPU first
            with deepspeed.zero.Init(config_dict_or_path=deepspeed_config()):
                model = cls(config, **kwargs)
        else:
            model = cls(config, **kwargs)
```
`is_deepspeed_zero3_enabled` from above snippet directly refers to the weakref in transformers.deepspeed
```python
def is_deepspeed_zero3_enabled():
    if _hf_deepspeed_config_weak_ref is not None and _hf_deepspeed_config_weak_ref() is not None:
        return _hf_deepspeed_config_weak_ref().is_zero3()
    else:
        return False
```

Due to above reasons I thought it would be good to let this be part of `transformers` repo as it is specifically used only in ZeRO Stage-3 for efficiently loading models (by directly partitioning params across GPUs) that are part of `transformers` repo.

[the remark this answers is not in the document]
Comment (sgugger) [reply src/accelerate/accelerator.py:186]: With the move, the weakref will disappear and we will rely on the `AcceleratorState` to know if zero3 is enabled inside Transformers. Again, not sure when is the right point to do the move (as it will make Accelerate a hard dep of Transformers) but want to flag this is the final destination :-)

File: docs/source/deepspeed.mdx [hunk 10]
```diff
@@ -0,0 +1,542 @@
...
+   In this situation, user has to use `accelerate.utils.DummyScheduler` to replace the PyTorch/Custom scheduler in their code. 
+
+   d. DS Optim + Custom Scheduler: The case when only `optimizer` key is present in the DeepSpeed config file. 
+   This will result in an error because one can only use DS Scheduler when using DS Optim.
+
+2. Notice the `auto` values in the above example DeepSpeed config files. 
+These have to specified by the user in the code by running `accelerator.state.deepspeed_plugin.deepspeed_config_process(**kwargs)`. 
+This allows user more flexibilty to fill these fields based on the code instead of hardcoding them. 
+Below is the snippet from `examples/by_feature/deepspeed_with_config_support.py` showing how to fill these: 
+```python
```
Comment (sgugger) [inline docs/source/deepspeed.mdx:387]: This is a huge block, I'd document less auto option plus make sure we automatically set the ones we can!

File: examples/by_feature/deepspeed_with_config_support.py [hunk 11]
```diff
@@ -0,0 +1,753 @@
...
+            assert extension in ["csv", "json", "txt"], "`validation_file` should be a csv, json or txt file."
+
+    if args.push_to_hub:
+        assert args.output_dir is not None, "Need an `output_dir` to create a repo when `--push_to_hub` is passed."
+
+    return args
+
+
+# New Code #
+def checkpoint_model(PATH, ckpt_id, model, epoch, last_global_step, **kwargs):
```
Comment (sgugger) [inline examples/by_feature/deepspeed_with_config_support.py:247]:
```suggestion
def checkpoint_model(checkpoint_folder, ckpt_id, model, epoch, last_global_step, **kwargs):
```
All-caps name are for constants, and here this is not a constant :-)

File: examples/by_feature/deepspeed_with_config_support.py [hunk 12]
```diff
@@ -0,0 +1,753 @@
...
+    """
+    checkpoint_state_dict = {
+        "epoch": epoch,
+        "last_global_step": last_global_step,
+    }
+    # Add extra kwargs too
+    checkpoint_state_dict.update(kwargs)
+
+    success = model.save_checkpoint(PATH, ckpt_id, checkpoint_state_dict)
+    status_msg = "checkpointing: PATH={}, ckpt_id={}".format(PATH, ckpt_id)
```
Comment (sgugger) [inline examples/by_feature/deepspeed_with_config_support.py:259]:
```suggestion
    success = model.save_checkpoint(checkpoint_folder, ckpt_id, checkpoint_state_dict)
    status_msg = f"checkpointing: checkpoint_folder={checkpoint_folder}, ckpt_id={ckpt_id}"
```
Variable name + f-strings only.

File: examples/by_feature/deepspeed_with_config_support.py [hunk 13]
```diff
@@ -0,0 +1,753 @@
...
+            optimizer=optimizer,
+            num_warmup_steps=args.num_warmup_steps,
+            num_training_steps=args.max_train_steps,
+        )
+    else:
+        lr_scheduler = DummyScheduler()
+
+    # New Code #
+    # Processing all the fields with `auto` values in the DeepSpeed Config file
+    if accelerator.state.deepspeed_plugin.config_file != "none":
```
Comment (sgugger) [inline examples/by_feature/deepspeed_with_config_support.py:559]: As discussed offline this is a very big specific block. We should try to avoid this at all cost.

File: src/accelerate/utils/deepspeed.py [hunk 14]
```diff
@@ -75,22 +51,65 @@ class DeepSpeedOptimizerWrapper(AcceleratedOptimizer):
...
+        self.param_groups = []
+        if params is None and param_groups is None:
+            raise ValueError("Either params or param_groups should be specified")
+        if param_groups is None:
+            self.param_groups.append({"params": params})
+        else:
+            self.param_groups = param_groups
+
+
+class DummyScheduler:
```
Comment (sgugger) [inline src/accelerate/utils/deepspeed.py:108]: Same here.

File: docs/source/deepspeed.mdx [hunk 15]
```diff
@@ -0,0 +1,542 @@
...
+main_process_port: null
+main_training_function: main
+mixed_precision: fp16
+num_machines: 1
+num_processes: 2
+use_cpu: false
+```
+
+```bash
+accelerate launch examples/by_feature/deepspeed_with_config_support.py \
```
[the remark this answers is not in the document]
Comment (pacman100) [reply docs/source/deepspeed.mdx]: Done.

File: docs/source/deepspeed.mdx [hunk 16]
```diff
@@ -0,0 +1,542 @@
...
+   this document is focused on this feature. This supports all the core features of DeepSpeed and gives user a lot of flexibility. 
+   User may have to change few lines of code depending on the config.
+2. Integration via `deepspeed_plugin`.This supports subset of the DeepSpeed features and uses default options for the rest of the configurations. 
+   User need not change any code and is good for those who are fine with most of the default settings of DeepSpeed.
+
+## What is integrated?
+
+Training:
+
+1. DeepSpeed ZeRO training supports the full ZeRO stages 1, 2 and 3 with ZeRO-Infinity (CPU and NVME offload).
```
[the remark this answers is not in the document]
Comment (pacman100) [reply docs/source/deepspeed.mdx]: Done.

File: docs/source/deepspeed.mdx [hunk 17]
```diff
@@ -0,0 +1,508 @@
...
+2. Integration via `deepspeed_plugin`.This supports subset of the DeepSpeed features and uses default options for the rest of the configurations. 
+   User need not change any code and is good for those who are fine with most of the default settings of DeepSpeed.
+
+## What is integrated?
+
+Training:
+
+1. DeepSpeed ZeRO training supports the full ZeRO stages 1, 2 and 3 as well as CPU/Disk offload of optimizer states, gradients and parameters. 
+Below is a short description of Data Parallelism using ZeRO - Zero Redundancy Optimizer along with diagram from this [blog post](https://www.microsoft.com/en-us/research/blog/zero-deepspeed-new-system-optimizations-enable-training-models-with-over-100-billion-parameters/)
+![ZeRO Data Parallelism](./imgs/parallelism-zero.png)
```
Comment (sgugger) [inline docs/source/deepspeed.mdx:46]: You should host the image in the `huggingface/documentation-images` dataset as we try to limit the size of the repo. Also those won't be copied when building the doc (the two images in this imgs folder are not used in the documentation, just the main README).

Comment (pacman100) [reply docs/source/deepspeed.mdx]: Done.

**Review by sgugger — APPROVED** [review]
LGTM, thanks a lot for working on this revamp!
